File: 31.Next-Permutation.go
// https://leetcode.com/problems/next-permutation/
//
// algorithms
// Medium (29.7%)
// Total Accepted:    198,530
// Total Submissions: 668,444
// beats 100.0% of golang submissions

package leetcode

import "sort"

func nextPermutation(nums []int) {
	length := len(nums)
	if length == 1 {
		return
	}
	if nums[length-1] > nums[length-2] {
		nums[length-1], nums[length-2] = nums[length-2], nums[length-1]
		return
	}

	idx := length - 1
	for idx > 0 {
		if nums[idx] > nums[idx-1] {
			break
		}
		idx--
	}
	if idx == 0 {
		sort.Sort(sort.IntSlice(nums))
	} else {
		minNum, minIdx := nums[idx], idx

		for i := idx + 1; i < length; i++ {
			if nums[i] < minNum && nums[i] > nums[idx-1] {
				minNum = nums[i]
				minIdx = i
			}
		}

		nums[minIdx], nums[idx-1] = nums[idx-1], nums[minIdx]
		sort.Sort(sort.IntSlice(nums[idx:]))
	}
}

File: 31.Next-Permutation.py
# https://leetcode.com/problems/next-permutation/
#
# algorithms
# Medium (30.69%)
# Total Accepted:    248,671
# Total Submissions: 810,285


class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        length = len(nums)
        if length < 2:
            return

        if length == 2:
            nums[0], nums[1] = nums[1], nums[0]
            return

        if nums[-1] > nums[-2]:
            nums[-1], nums[-2] = nums[-2], nums[-1]
            return

        i = length - 1
        while i > 0 and nums[i - 1] >= nums[i]:
            i -= 1

        if i == 0:
            nums.sort()
            return

        swap_idx, delta = i, float('inf')
        for j in xrange(i, length):
            if nums[j] > nums[i - 1] and nums[j] - nums[i - 1] < delta:
                swap_idx = j
        nums[swap_idx], nums[i - 1] = nums[i - 1], nums[swap_idx]

        tmp = nums[i:]
        tmp.sort()
        for j in xrange(i, length):
            nums[j] = tmp[j - i]

        return
